### PR TITLE
Don't drop content-specific headers on requests without a body

### LINKED ISSUE #618: Investigate content headers being dropped when content-length is 0
In theory you can have a POST request with a Content-Length: 0, but still have other content headers like Content-Type which would be used for routing the request at the destination. 

In today's HttpProxy logic we avoid creating the HttpContent on the request if there's no body since HttpClient will automatically create a chunked body for any HttpContent without a specified length (and a servers treat requests like GETs with a chunked body as invalid). We may need to special case zero length bodies and provide an empty content object so that we can still forward headers like Content-Type (would HttpClient allow that?).

### PR DESCRIPTION
**Pull request by MihaZupan**
Fixes #618

Headers defined as `HttpHeaderType.Content` in [KnownHeaders.cs](https://github.com/dotnet/runtime/blob/c90debb5dc8ee85840bbef1a44b8502ab169f47b/src/libraries/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs#L28-L61) may only be present in the `HttpContentHeaders` collection.
To avoid dropping such headers when there is no request body (e.g. `Content-Length` is 0), we inject a dummy 0-length content object to hold the headers.

### CODE REVIEW

File: test/ReverseProxy.Tests/Service/Proxy/HttpProxyTests.cs [hunk 1]
```diff
@@ -448,42 +448,72 @@ public async Task ProxyAsync_UpgradableRequestFailsToUpgrade_ProxiesResponse()
...
                 {
                     Assert.Equal(new Version(2, 0), request.Version);
                     Assert.Equal(method, request.Method.Method, StringComparer.OrdinalIgnoreCase);
 
-                    Assert.Null(request.Content);
+                    // When adding content-specific headers, we will inject an EmptyHttpContent if no other content is present
+                    if (headers.Any())
+                    {
+                        Assert.NotNull(request.Content);
+                        Assert.Equal("EmptyHttpContent", request.Content.GetType().Name);
```
Comment (alnikola) [inline test/ReverseProxy.Tests/Service/Proxy/HttpProxyTests.cs:498]: Could we compare actual types here?
```suggestion
                        Assert.Equal(typeof(EmptyHttpContent), request.Content.GetType());
```

File: src/ReverseProxy/Utilities/RequestUtilities.cs [hunk 2]
```diff
@@ -101,23 +123,38 @@ internal static void AddHeader(HttpRequestMessage request, string headerName, St
                 string headerValue = value;
                 if (!request.Headers.TryAddWithoutValidation(headerName, headerValue))
                 {
+                    if (request.Content is null && _contentHeaders.Contains(headerName))
+                    {
+                        request.Content = new EmptyHttpContent();
```
Comment (Kahbazi) [inline src/ReverseProxy/Utilities/RequestUtilities.cs:128]: Is it possible to stash one instance of 
`EmptyHttpContent` and reuse that?

Comment (MihaZupan) [reply src/ReverseProxy/Utilities/RequestUtilities.cs:128]: It serves as a container for `HttpContentHeaders` that may change between requests

**Review by alnikola — APPROVED** [review]
LGTM

**Review by Tratcher — APPROVED** [review]

**Pull request merged**